### PR TITLE
Few fixes and improvements

### PR DESCRIPTION
**Pull request by vlasenkoalexey**
I'm using monotorrent in my App, here are some changes I have made to the latest version.

### CODE REVIEW

**Comment by alanmcgovern** [comment]
Hi! Some of these changes look quite useful. I'd love to include them and there's one thing you could do to make it easier for me to do so!

In general if one pull request is created per feature/changeset it makes it easier to review, iterate and merge changes. This PR has a single commit with many different changesets This means I have a choice of accepting everything or nothing. I'll annotate the PR with thoughts on how things could be split up for submission. 

Thanks!

File: src/MonoTorrent/MonoTorrent.Client.Connections/SocketConnection.cs [hunk 1]
```diff
@@ -198,16 +198,20 @@ public ReusableTask<int> ReceiveAsync (byte[] buffer, int offset, int count)
             args.SetBuffer (offset, count);
             args.UserToken = ReceiveTcs;
 
+            // https://github.com/alanmcgovern/monotorrent/issues/202
+#if DEBUG
             AsyncFlowControl? control = null;
             if (!ExecutionContext.IsFlowSuppressed ())
                 control = ExecutionContext.SuppressFlow ();
-
+#endif
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client.Connections/SocketConnection.cs]: If this is wrapped in `#if DEBUG` then it will only have an effect when running local debug builds. The NuGet packages from nuget.org are built with the `Release` configuration so this code will still execute. Is this intended?

Also, this changeset would be good as a separate PR.

File: src/MonoTorrent/MonoTorrent.Client.Messages/MessageBundle.cs [hunk 2]
```diff
@@ -85,5 +86,10 @@ public override int Encode (byte[] buffer, int offset)
 
             return CheckWritten (written - offset);
         }
+
+        public override string ToString()
+        {
+            return $"MessageBundle [{String.Join(", ", this.Messages.Select(m => m.ToString()).ToArray())}]";
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client.Messages/MessageBundle.cs]: Do you want `ToString` to aid debugging? If so you could implement it similar to how BitField does it ( https://github.com/alanmcgovern/monotorrent/blob/master/src/MonoTorrent/MonoTorrent/BitField.cs#L43 )

File: src/MonoTorrent/MonoTorrent.Client.Messages/PeerMessage.cs [hunk 3]
```diff
@@ -107,5 +107,10 @@ internal void Handle (TorrentManager manager, PeerId id)
         {
             manager.Mode.HandleMessage (id, this);
         }
+
+        public override string ToString()
+        {
+            return this.GetType().Name;
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client.Messages/PeerMessage.cs]: The default ToString implementation should return the typename. https://docs.microsoft.com/en-us/dotnet/api/system.object.tostring?view=netframework-4.8#remarks . That should mean this change is unnecessary?

File: src/MonoTorrent/MonoTorrent.Client.Tracker/TrackerTier.cs [hunk 4]
```diff
@@ -36,7 +36,6 @@ public class TrackerTier
     {
         public IList<ITracker> Trackers { get; }
 
-        internal bool SendingStartedEvent { get; set; }
         internal bool SentStartedEvent { get; set; }
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client.Tracker/TrackerTier.cs]: :+1:

File: src/MonoTorrent/MonoTorrent.Client.Tracker/UdpTracker.cs [hunk 5]
```diff
@@ -60,6 +60,8 @@ protected override async Task<List<Peer>> DoAnnounceAsync (AnnounceParameters pa
                 var message = new AnnounceMessage (DateTime.Now.GetHashCode (), connectionId, parameters);
                 var announce = (AnnounceResponseMessage) await SendAndReceiveAsync (message);
                 MinUpdateInterval = announce.Interval;
+                Complete = announce.Seeders;
+                Incomplete = announce.Leechers;
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client.Tracker/UdpTracker.cs]: This would be good as a separate PR, ideally with a unit test. I can add one once the change is split up if you do not have the time to do so!

File: src/MonoTorrent/MonoTorrent.Client/ClientEngine.cs [hunk 6]
```diff
@@ -77,6 +78,9 @@ public class ClientEngine : IDisposable
         public event EventHandler<StatsUpdateEventArgs> StatsUpdate;
         public event EventHandler<CriticalExceptionEventArgs> CriticalException;
 
+        public event EventHandler<TorrentEventArgs> TorrentRegistered;
+        public event EventHandler<TorrentEventArgs> TorrentUnregistered;
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client/ClientEngine.cs]: This looks like a good addition. Could this, and related changes, be added in a separate PR?

File: src/MonoTorrent/MonoTorrent.Client/ClientEngine.cs [hunk 7]
```diff
@@ -468,7 +475,9 @@ private void LogicTick ()
                 uploadLimiter.UpdateChunks (Settings.MaximumUploadSpeed, TotalUploadSpeed);
             }
 
-            ConnectionManager.CancelPendingConnects ();
+            for (int i = 0; i < this.torrents.Count; i++)
+                ConnectionManager.CancelPendingConnects(this.torrents[i]);
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client/ClientEngine.cs]: This change looks like it's broken. It should cause pending connects to be closed every time LogicTick is invoked rather than only when the timeout has been reached. This means pending connections will be closed between 0 and 500ms after the connection attempt is initiated. Do you recall what issue you were trying to solve with this change?

File: src/MonoTorrent/MonoTorrent.Client/Managers/TorrentManager.cs [hunk 8]
```diff
@@ -286,6 +288,48 @@ public class TorrentManager : IDisposable, IEquatable<TorrentManager>
...
         public InfoHash InfoHash { get; }
 
+        /// <summary>
+        /// Torrent tracker announce Urls
+        /// </summary>
+        public IList<RawTrackerTier> AnnounceUrls
+        {
+            get;
+            private set;
+        }
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client/Managers/TorrentManager.cs]: This shouldn't be exposed at this level. Why did you need access to this here?

File: src/MonoTorrent/MonoTorrent.Client/Managers/TorrentManager.cs [hunk 9]
```diff
@@ -406,7 +451,7 @@ void CreateRateLimiters ()
 
         #region Public Methods
 
-        internal void ChangePicker (PiecePicker picker)
+        public void ChangePicker(PiecePicker picker)
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client/Managers/TorrentManager.cs]: This should not be public as it is not threadsafe for users of the library to invoke this directly. A threadsafe method does exist (`ChangePickerAsync`) so that's what people should be using.

File: src/MonoTorrent/MonoTorrent.Client/Managers/TrackerManager.cs [hunk 10]
```diff
@@ -139,6 +139,10 @@ async Task Announce (TorrentEvent clientEvent, ITracker referenceTracker)
                         actualArgs = actualArgs.WithClientEvent (TorrentEvent.Started);
 
                     var peers = await tuple.Item2.AnnounceAsync (actualArgs);
+                    if (clientEvent == TorrentEvent.Started)
+                    {
+                        tuple.Item1.SentStartedEvent = true;
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client/Managers/TrackerManager.cs]: This looks good! It could be it's own PR too.

File: src/MonoTorrent/MonoTorrent.Client/Managers/TorrentManager.cs [hunk 11]
```diff
@@ -286,6 +288,48 @@ public class TorrentManager : IDisposable, IEquatable<TorrentManager>
...
+        /// <summary>
+        /// Magnet link
+        /// </summary>
+        public MagnetLink MagnetLink {
+            get {
+                if (this.magnetLink == null) {
+                    this.magnetLink = new MagnetLink (Torrent.InfoHash, name: Name,
+                        announceUrls: AnnounceUrls != null ? AnnounceUrls.SelectMany (rawTrackerTier => rawTrackerTier).ToList () : new List<string> ());
+                }
+                return this.magnetLink;
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client/Managers/TorrentManager.cs]: This set of changes looks good and would be ideal as it's own PR!

File: src/MonoTorrent/MonoTorrent.Client/Managers/TorrentManager.cs [hunk 12]
```diff
@@ -286,6 +288,48 @@ public class TorrentManager : IDisposable, IEquatable<TorrentManager>
...
+            private set;
+        }
+
+        /// <summary>
+        /// Magnet link
+        /// </summary>
+        public MagnetLink MagnetLink {
+            get {
+                if (this.magnetLink == null) {
+                    this.magnetLink = new MagnetLink (Torrent.InfoHash, name: Name,
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.Client/Managers/TorrentManager.cs]: Couldn't we just invoke `Torrent.ToMagnetLink` rather than reimplementing it? This method does rely on `Torrent` being non-null so it should simpler!

**Comment by vlasenkoalexey** [comment]
sg, I'll will split this PR into separate ones for each specific change one I have some time.

File: src/MonoTorrent/MonoTorrent.Client.Connections/SocketConnection.cs (hunk 1 above)
Comment (alanmcgovern) [reply src/MonoTorrent/MonoTorrent.Client.Connections/SocketConnection.cs]: A similar change has been integrated and released in the latest nuget package. There is a bug in some older versions of UWP which causes nullreference exceptions if you use this API. I assume you encountered that issue :)

File: src/MonoTorrent/MonoTorrent.Client.Tracker/TrackerTier.cs (hunk 4 above)
Comment (alanmcgovern) [reply src/MonoTorrent/MonoTorrent.Client.Tracker/TrackerTier.cs]: I removed this variable :)
